checkisjson.go

```diff
@@ -0,0 +1,153 @@
+		// Even if there's no media type, we want to see something useful.
+		mediaType = fmt.Sprintf("%q", contentType)
+	}
+	// TODO use charset.NewReader to convert from non-utf8 content?
+	// Read the body ignoring any errors - we'll just make do with what we've got.
+	bodyData, _ := ioutil.ReadAll(io.LimitReader(noErrorReader{body}, int64(maxErrorBodySize)))
+	switch mediaType {
+	case "text/html":
+		text, err := htmlToText(bytes.NewReader(bodyData))
+		if err != nil {
+			// Note: it seems that this can never actually
+			// happen - the only way that the HTML parser
+			// can fail is if there's a read error and we've
+			// removed that possibility by using
+			// noErrorReader above.
+			return errgo.Notef(err, "unexpected (and invalid) content text/html; want application/json; content: %q", sizeLimit(bodyData))
+		}
+		if len(text) == 0 {
+			return errgo.Newf(`unexpected content type text/html; want application/json; content: %q`, sizeLimit(bodyData))
+		}
+		return errgo.Newf(`unexpected content type text/html; want application/json; content: %s`, sizeLimit(text))
+	case "text/plain":
+		return errgo.Newf(`unexpected content type text/plain; want application/json; content: %s`, sizeLimit(saneText(string(bodyData), true)))
+	default:
+		return errgo.Newf(`unexpected content type %s; want application/json; content: %q`, mediaType, sizeLimit(bodyData))
+	}
+}
+
+type noErrorReader struct {
+	r io.Reader
+}
+
+func (r noErrorReader) Read(buf []byte) (int, error) {
+	n, err := r.r.Read(buf)
+	if err != nil {
+		err = io.EOF
+	}
+	return n, err
+}
+
+func sizeLimit(data []byte) []byte {
+	const max = 1024
+	if len(data) < max {
+		return data
+	}
+	return append(data[0:max], fmt.Sprintf(" ... [%d bytes omitted]", len(data)-max)...)
+}
+
+// htmlToText attempts to return some relevant textual content
+// from the HTML content in the given reader, formatted
+// as a single line.
+func htmlToText(r io.Reader) ([]byte, error) {
+	n, err := html.Parse(r)
+	if err != nil {
+		return nil, err
+	}
+	var buf bytes.Buffer
+	htmlNodeToText(&buf, n)
+	return buf.Bytes(), nil
+}
+
+func htmlNodeToText(w *bytes.Buffer, n *html.Node) {
+	for ; n != nil; n = n.NextSibling {
+		switch n.Type {
+		case html.TextNode:
+			data := saneText(n.Data, false)
+			if len(data) == 0 {
+				break
+			}
+			if w.Len() > 0 {
+				w.WriteString("; ")
+			}
+			w.Write(data)
+		case html.ElementNode:
+			if n.DataAtom != atom.Script {
+				htmlNodeToText(w, n.FirstChild)
+			}
+		case html.DocumentNode:
+			htmlNodeToText(w, n.FirstChild)
+		}
+	}
+}
+
+// saneText tries to make the given string easier to read when presented
+// as a single line. It squashes each run of white space into a single
+// space, trims leading and trailing white space and trailing full
+// stops. If newlineSemi is true, any newlines will be replaced with a
+// semicolon.
+func saneText(s string, newlineSemi bool) []byte {
+	out := make([]byte, 0, len(s))
+	prevWhite := false
+	for _, r := range s {
```

Comment:
What does this do if the content isn't utf-8? I answered my own question http://play.golang.org/p/Hm4qAlwUvu it uses the replacement character, so its all good.